### DISCUSSION TITLE
Equivalent to witness and assembly

### DISCUSSION THREAD

**Discussion by bkreiser** [bloxbean/cardano-client-lib#41] [Q&A] [answered]
Does cardano-client-lib handle witness and assembly?  Those are capabilities of the cardano-cli providing the ability to distribute the transaction to other parties and having them sign it and then return a file that you can then use in assemble.

**Comment by satran004** [comment, marked as answer]
Hi,

It's possible. Please go through the below options using low-level and high-level apis. Let me know if there is any gap. 

**Option 1: Using Low-level serialization api**

a> Using low level api, you can create a Transaction object from TransactionInput[] and TransactionOutput[]. You need to find the required utxos through UtxoService / UtxoSelectionStrategy impl.

After building the transaction, you can sign the transaction sequentially or  distribute and then assemble.

b> Sign the transaction with account1
c> Sign the output from step-b with account2. This will add additional witness
d> Post transaction with output from step-c

Check this example (Transfer fund from multi-sig script address, at least 2 witnesses): https://github.com/bloxbean/cardano-client-examples/blob/5978765bac528adca9e981e8f3fad0b892abd232/src/main/java/com/bloxbean/cardano/client/examples/TransferFromMultisigScriptAddress.java#L115

Or,

b1> Account 1 sign the original transaction
c1> Account 2 sign the original transaction
d1> Get witness from ouput from step-c1 , add to original transaction.
e1> Get witness from ouput from step-d1 , add to original transaction. 
f1> Submit the original transaction
https://github.com/bloxbean/cardano-client-examples/blob/5978765bac528adca9e981e8f3fad0b892abd232/src/main/java/com/bloxbean/cardano/client/examples/TransferFromMultisigScriptAddress.java#L118

**Option 2: Using High Level api with some tricks**

Check the same example but using PaymentTransaction api : - https://github.com/bloxbean/cardano-client-examples/blob/5978765bac528adca9e981e8f3fad0b892abd232/src/main/java/com/bloxbean/cardano/client/examples/MultiWitnessWithPaymentTransaction.java#L34

Using PaymentTransaction api, it's possible to achieve the same. But some workaround required, like creating a ReadOnlyAccount class to pass  sender to PaymentTransaction. Because sender here is a script address.
Currently there is no way to add NativeScript as witness directly to PaymentTransaction (A possible enhancement for future release)

- Use UtxoTransactionBuilder to build Transaction object from PaymentTransaction
- Once you get the Transaction object, rest of the steps are similar to Option-1

Probably we can update PaymentTransaction API to support multisig transaction without above workarounds.

**Reply by bkreiser** [reply]
Thanks for the rich response.  I will let you know how it goes. :)

**Reply by edridudi** [reply]
I am trying to perform the same while account1 = nami wallet at client's side.
It is Possible that nami user will sign the transaction only once.
thus trying to figure a way to calculate fees before Transaction sign phase.

**Reply by satran004** [reply]
> I am trying to perform the same while account1 = nami wallet at client's side. It is Possible that nami user will sign the transaction only once. thus trying to figure a way to calculate fees before Transaction sign phase.

Only difference in a signed transaction object is witnesses in transactionWitness list. If you can create fake witness just for fee calculation, it should be ok.

I think  similar approach is currently done by cardano-serialization-lib rust library 
https://github.com/Emurgo/cardano-serialization-lib/blob/fd37ca5f4d888aca4ba06fa40f996373474e38b3/rust/src/tx_builder.rs#L43

Probably, I should do the same in cardano-client-lib to avoid double signing.

**Reply by edridudi** [reply]
> > I am trying to perform the same while account1 = nami wallet at client's side. It is Possible that nami user will sign the transaction only once. thus trying to figure a way to calculate fees before Transaction sign phase.
> 
> Only difference in a signed transaction object is witnesses in transactionWitness list. If you can create fake witness just for fee calculation, it should be ok.
> 
> I think similar approach is currently done by cardano-serialization-lib rust library https://github.com/Emurgo/cardano-serialization-lib/blob/fd37ca5f4d888aca4ba06fa40f996373474e38b3/rust/src/tx_builder.rs#L43
> 
> Probably, I should do the same in cardano-client-lib to avoid double signing.

That's exactly what I did since I wrote my comment.
I Used Fake transactionWitnesses just to calculate the Fee.
I also use the Multi-Sig Transaction to mint a token for the Nami wallet user so all I have left is to handle this error:

Result{successful=false, response='{"error":"Bad Request","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraAlonzo (ApplyTxError [UtxowFailure (WrappedShelleyEraFailure (ScriptWitnessNotValidatingUTXOW (fromList [ScriptHash \\\"654ebfc69ea9b582d09755a0760fdac7b3e16718ef47acd958708035\\\"])))])\"","status_code":400}', code=400, value=null}

I will share my Code with this repo after I'll manage to understand the problem and fix it.
